### ISSUE TITLE
Prepared statements aren't invalidated after schema change

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#2454]
Example test that fails when `cachePreparedStatements` is true and passes when it's false:
```dart
import 'package:drift/native.dart';
import 'package:drift/drift.dart';
import 'package:test/test.dart';

void main() {
  driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  for (final cachePreparedStatements in [true, false]) {
    test('cachePreparedStatements=$cachePreparedStatements', () async {
      final db = TestDb(cachePreparedStatements);
      await db.customStatement('create table t (c1)');
      await db.customInsert('insert into t values (1)');
      final before = await db.customSelect('select * from t').getSingle();
      expect(before.data, {'c1':1});
      await db.customStatement('alter table t add column c2');
      final after = await db.customSelect('select * from t').getSingle();
      expect(after.data, {'c1':1, 'c2':null});
    });
  }
}

class TestDb extends GeneratedDatabase {
  TestDb(bool cachePreparedStatements) : super(
    NativeDatabase.memory(cachePreparedStatements: cachePreparedStatements)
  );
  @override final List<TableInfo> allTables = const [];
  @override final int schemaVersion = 1;
}
```

For context, we have some custom migration test tooling (that predates Drift's migration test APIs) where each migration must have an associated test. The tests roughly take the form of: prepare a database with a known schema and set of data, run the migration, then run some assertions that the migration succeeded. Any test which ran the same query before and after the migration started failing after https://github.com/simolus3/drift/pull/2434.

In our case we only encountered this in test code because we're still using the `sqflite` backend in production, and (as you noted in another issue recently) drift is not optimized for a dynamic schema so this probably isn't super common to hit, but it's probably still possible to write a `MigrationStrategy` that exhibits the bug.

**Comment by Mike278** [comment]
Actually after looking into this a bit more, this is kind of surprising because the [sqlite docs](https://www.sqlite.org/rescode.html#schema) indicate that "If a prepared statement is generated from sqlite3_prepare_v2() then the statement is automatically re-prepared if the schema changes", yet this test fails:
```dart
test('sqlite3 direct', () {
  final db = sqlite3.openInMemory()
    ..execute('create table t (c1)')
    ..execute('insert into t values (1)');
  final stmt = db.prepare('select * from t');
  expect(stmt.select(), [{'c1': 1}]);
  db.execute('alter table t add column c2');
  expect(stmt.select(), [{'c1': 1, 'c2': null}]);
});
```

I tried debugging a bit but couldn't find anything obvious - [this](https://github.com/simolus3/sqlite3.dart/blob/c16721003bffa6e9276ba9de8983c3f98e041b40/sqlite3/lib/src/implementation/statement.dart#LL52C35-L52C55) call to `sqlite3_column_count` surprisingly returns 1 both times. 

(I'm guessing the sqlite docs are slightly outdated because while it does say "generated from sqlite3_prepare_v2", I get the same behaviour with sqlite3_prepare_v3, and I never see any `SQLITE_SCHEMA` return codes.)

**Comment by simolus3** [comment]
Good catch! The issue is that, while statements are re-compiled in response to schema changes, this happens in `sqlite3_step()`. In the `sqlite3` package, we're first collecting column names and then step through the statement to convert rows into a map-like structure. So if a statement gets recompiled, we're using the wrong columns.

I think this can be fixed in the sqlite3 package by calling `step` first, but I need to investigate some corner cases (like what happens if the schema changes while we're iterating through a statement's cursor).

And luckily there's no security risk when the amount of columns in a statement decreases due to a recompilation, sqlite returns null for out-of-bounds column reads.

**Comment by simolus3** [comment]
Fixed in version `1.11.2` of the `sqlite3` package.

**Comment by Mike278** [comment]
Ah that's tricky, nice fix!
